### PR TITLE
Refactor array compaction functions

### PR DESCRIPTION
**Pull request by harendra-kumar**

### CODE REVIEW

File: core/src/Streamly/Internal/Data/MutArray/Type.hs [hunk 1]
```diff
@@ -817,6 +814,10 @@ reallocExplicit elemSize newCapacityInBytes MutArray{..} = do
         , arrBound = newCapInBytes
         }
 
+-- XXX Should these be called reallocBytes etc?
+-- XXX We may also need reallocAs to allocate as pinned/unpinned explicitly. In
+-- fact clone/pinnedClone can be implemented using reallocAs.
```
Comment (adithyaov) [inline core/src/Streamly/Internal/Data/MutArray/Type.hs:819]: Can you open an issue?

File: core/src/Streamly/Internal/Data/MutArray/Type.hs [hunk 2]
```diff
@@ -1114,7 +1119,7 @@ getIndicesWith = indexReaderWith
 -- /Pre-release/
 {-# INLINE indexReader #-}
 indexReader :: (MonadIO m, Unbox a) => Stream m Int -> Unfold m (MutArray a) a
-indexReader = getIndicesWith liftIO
```
Comment (adithyaov) [inline core/src/Streamly/Internal/Data/MutArray/Type.hs:1117]: Why was there no warning here?

**Review by adithyaov — APPROVED** [review]

**Review by adithyaov — APPROVED** [review]

File: core/src/Streamly/Internal/Data/MutArray/Type.hs [hunk 3]
```diff
@@ -2737,44 +2760,45 @@ rCompactLE n (D.Stream step state) =
...
+               in if len >= minBytes
                   then FL.Done arr
                   else FL.Partial (Just arr)
+    -- XXX Buffer arrays as a list to avoid copy and reallocations
     step (Just buf) arr = do
         let len = byteLength buf + byteLength arr
         buf1 <-
             if byteCapacity buf < len
-            then liftIO $ realloc (max len nBytes) buf
+            then liftIO $ realloc (max len minBytes) buf
```
Comment (harendra-kumar) [inline core/src/Streamly/Internal/Data/MutArray/Type.hs:2798]: remove liftIO.

File: core/src/Streamly/Internal/Data/MutArray/Type.hs [hunk 4]
```diff
@@ -2786,59 +2810,54 @@ fCompactGE n = Fold step initial extract extract
...
+            else return $ FL.Partial $ Tuple' (Just buf) acc
+
+    step (Tuple' Nothing r1) arr =
+         runInner (byteLength arr) r1 arr
+
+    -- XXX Buffer arrays as a list to avoid copy and reallocations
+    step (Tuple' (Just buf) r1) arr = do
+        let len = byteLength buf + byteLength arr
+        buf1 <- if byteCapacity buf < len
+                then liftIO $ realloc (max minBytes len) buf
```
Comment (harendra-kumar) [inline core/src/Streamly/Internal/Data/MutArray/Type.hs:2854]: remove liftIO.

**Pull request merged**